### ISSUE TITLE
Cloud storage: Should always upload entire ruleset, ignoring current filter

### ISSUE THREAD

**Issue by talklittle** [uBlockOrigin/uMatrix-issues#27]
### Prerequisites

* [x] Did you perform a cursory search of the issue tracker?
    
* Can you reproduce the problem when...
    * [ ] uMatrix is the only extension?
    * [ ] uMatrix with default lists/settings?
    * [ ] using a new, unmodified browser profile?
* [x] Are you running the latest version?
* [x] Did you check the [documentation](https://github.com/gorhill/uMatrix/wiki)?

### Description

On the "My rules" tab, if you have typed a filter into the CodeMirror rule filter input box, and then press the Upload button for cloud storage, then uMatrix only uploads the currently visible rules. This is the most obvious case causing issue.

In some other cases that are more difficult to reproduce, when adding new rules, even without having a filter typed in the input box, uMatrix occasionally only uploads the latest diff(?), resulting in lost rules.

The ideal behavior is to always upload the full ruleset to cloud storage, regardless of filters and recent changes.

### A specific URL where the issue occurs

moz-extension://.../dashboard.html#user-rules

### Steps to Reproduce

1. Enable cloud storage
2. Go to My rules in uMatrix
3. Type a filter
4. Press Upload button for cloud storage
5. Clear filter
6. Press Download button for cloud storage
7. See that only filtered subset of rules remains

### Supporting evidence

![umatrix-cloud-storage-upload-filtered](https://user-images.githubusercontent.com/87961/41985184-7a1f15e2-79e7-11e8-98a4-db3cbfb89e33.gif)

### Your environment

* uMatrix version: 1.3.10
* Browser Name and version: Firefox 60.0.2
* Operating System and version: Ubuntu 16.04.4

**Comment by uBlock-user** [comment]
Revert button is there to revert it immediately, it doesn't saved, so you're not losing any rules.

**Comment by talklittle** [comment]
But the cloud backup is indeed lost...? Unless you are in the habit of testing the backup *every. single. time.* you press the Upload button. Which is only necessary in the case that uMatrix's backup functionality cannot be trusted to work as expected.

**Comment by uBlock-user** [comment]
> But the cloud backup is indeed lost...?


> Secondly, after clicking Revert, click upload again and your original set of rules will be uploaded again, overwriting those temp rules.

**Comment by talklittle** [comment]
> Unless you are in the habit of testing the backup every. single. time. you press the Upload button.

**Comment by uBlock-user** [comment]
The thing is this kind of accident can never really happen in the first place. You using a filter and then accidently clicking the far above cloud upload button is not common doesn't happen just like that and won't happen unless you're experimenting or don't have any idea what that button means. 

If I were to suggest, it would be to only upload Permanent Rules, so no matter what changes you make to Temp Rules or even if you accidently delete everything in Temp rules, only backups that are created would be the ones you save in Permanent Rules.

**Comment by talklittle** [comment]
> If I were to suggest, it would be to only upload Permanent Rules

Agreed. But that's kind of an orthogonal issue to the Filter which causes the reported problem.

> You using a filter and then accidently clicking the far above cloud upload button is not common doesn't happen just like that and won't happen unless you're experimenting or don't have any idea what that button means.

Please explain a use case for the cloud storage uploading the current filtered view only. I could *maybe* understand the "Export to file" button working that way for very specific use cases, but not cloud storage backup.

**Comment by gorhill** [comment]
This is not the behavior I intended, this is a regression.

**Comment by uBlock-user** [comment]
> Please explain a use case for the cloud storage uploading the current filtered view only

Not sure what you mean here. Lets say I'm searching for mega.nz rule, I add it to the filter mega.nz and the temp rules will only display mega.nz rule, I will do my business with it(edit/delete/add) and then either click Commit or Revert if I'm unhappy. That's it. I won't just accidently or immediately click on Cloud Storage, that won't happen. 

I'm not saying your suggestion is wrong by any means, but that Revert button is there to correct the wrong if it ever happens and it won't happen just like that, so chances of it happening are slim, so using revert and then clicking Upload undoes the lost backup, however if I were to make a suggesiton still, it would be that only upload Permannent Rules to the storage as they only matter.

**Comment by talklittle** [comment]
Okay, here's my use case. Say I added a new rule, e.g. for twitter.com, using the overlay panel and not My rules. I want to make sure it's listed in My rules so I open My rules, type the filter "twitter" and see that, yes, the rule was successfully added to My rules and I'm happy. So I hit Upload.

**Comment by uBlock-user** [comment]
Sure then it can happen, and for that to not happen, you would have to be cautious not to click on Upload button when using a filter to check for any rules and only click once you remove the filter and the original rules are back, atleast that's how I do it, but anyways this discussion no longer matters because I thought this was _by-design_, which is not the case, so the whole argument was based on that scenario.
